src/txmempool.cpp

```diff
@@ -151,33 +152,47 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
     }
 }
 
-bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntries &setAncestors, uint64_t limitAncestorCount, uint64_t limitAncestorSize, uint64_t limitDescendantCount, uint64_t limitDescendantSize, std::string &errString, bool fSearchForParents /* = true */) const
+bool CTxMemPool::CalculateMemPoolAncestors(const std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef>& entries,
```

Comment:
I think that `CalculateMemPoolAncestors()` was already doing too much, and this adds just a little bit more complication. Instead of overloading CMPA and having CMPA(tx) call CMPA(package), what do you think about the splitting it into two parts:

- calculate parents (two different functions - one for a single transaction which may have fSearchForParents=false, and one for a package)
- check package limits (everything from `while (!staged_ancestors.empty())` downwards, which is the same function for an individual transaction and a package if parametrized correctly)

and then roughly:

```
CMPA(tx) {
    set_ancestors = CalculateParents(tx);
    CheckPackageLimits(set_ancestors);
}

CMPA(package) {
    set_ancestors = CalculateParents(package);
    CheckPackageLimits(set_ancestors);
}
```

Reply:
nit: use function signatures with spans when passing a const vec, removes the need to allocate for passing a single tx (can span without allocating from pointer).

Reply:
re: @jnewbery's feedback to split, I think that might be right if fsearchforparents should never be used as false with a package? the asserts confuse me...